### PR TITLE
Migrate `output-type-permutations` `run-make` test to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

try-job: x86_64-msvc
try-job: armhf-gnu
try-job: aarch64-apple
try-job: test-various

### CODE REVIEW

File: tests/run-make/output-type-permutations/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    fs_wrapper::remove_file("foo");
+    rustc().input("foo.rs").crate_type("rlib").emit("link=foo").run();
+    fs_wrapper::remove_file("foo");
+    rustc().input("foo.rs").crate_type("rlib").arg("--emit=link=foo").run();
+    fs_wrapper::remove_file("foo");
+    assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);
+
+    rustc().input("foo.rs").crate_type("dylib").emit("link").output("foo").run();
+    fs_wrapper::remove_file("foo");
+    rustc().input("foo.rs").crate_type("dylib").emit("link=foo").run();
```
Comment (Kobzol) [inline tests/run-make/output-type-permutations/rmake.rs:100]: This was `bin_foo` in the original.

File: tests/run-make/output-type-permutations/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    fs_wrapper::remove_file("foo");
+    rustc().input("foo.rs").crate_type("dylib").arg("--emit=link=foo").run();
+    fs_wrapper::remove_file("foo");
+    remove_artifacts();
+    assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);
+
+    rustc().input("foo.rs").crate_type("staticlib").emit("link").output("foo").run();
```
Comment (Kobzol) [inline tests/run-make/output-type-permutations/rmake.rs:107]: And here.

File: tests/run-make/output-type-permutations/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    fs_wrapper::remove_file("foo");
+    rustc().input("foo.rs").crate_type("staticlib").emit("link=foo").run();
+    fs_wrapper::remove_file("foo");
+    rustc().input("foo.rs").crate_type("staticlib").arg("--emit=link=foo").run();
+    fs_wrapper::remove_file("foo");
+    assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);
+
+    let bin_foo = bin_name("foo");
```
Comment (Kobzol) [inline tests/run-make/output-type-permutations/rmake.rs:115]: This was already defined above.

File: tests/run-make/output-type-permutations/rmake.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    rustc().input("foo.rs").crate_type("bin").emit("link").output(&bin_foo).run();
```
Comment (Kobzol) [inline tests/run-make/output-type-permutations/rmake.rs:116]: And here.

File: tests/run-make/output-type-permutations/rmake.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    fs_wrapper::remove_file("ir");
+    fs_wrapper::remove_file("bc");
+    fs_wrapper::remove_file("obj");
+    fs_wrapper::remove_file("link");
+    assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);
+
+    rustc()
+        .input("foo.rs")
+        .arg("--emit=asm=asm")
+        .emit("llvm-ir=ir")
```
Comment (Kobzol) [inline tests/run-make/output-type-permutations/rmake.rs:149]: I would use manual `arg` here, to make it clear that the difference is between `emit x` and `emit=x`.

**Review by Kobzol — COMMENTED** [review]
Well, what a *fun* test this is. I think that the "middle section" (at least lines 48 - 78, but maybe even until line 123) could be made more readable by introducing some helper, that would receive `rustc` in a closure, set flags for it, fill in the emit combinations, and remove the created files between each invocation. But maybe the test structure is too irregular for that, it would need to be tried :)

File: tests/run-make/output-type-permutations/rmake.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    std::fs::remove_file("bar.exe.a").unwrap_or_default();
+}
+
+fn main() {
+    rustc().input("foo.rs").crate_type("rlib,dylib,staticlib").run();
+    fs_wrapper::remove_file(rust_lib_name("bar"));
+    fs_wrapper::remove_file(dynamic_lib_name("bar"));
+    fs_wrapper::remove_file(static_lib_name("bar"));
+    remove_artifacts();
+    assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);
```
Comment (jieyouxu) [inline tests/run-make/output-type-permutations/rmake.rs:32]: Suggestion: I want to say this check in itself is not sufficient, as my reading of the test comment we want to be checking two conditions here:

(1) unexpected files are not created
(2) expected files are created

Checks like `assert_eq!(fs_wrapper::read_dir(cwd()).count(), 1);` are only a *proxy* for checking for the correct *number* of files, but it does not actually check what the precise files are.

I would imagine the test looking something like

```rs
fn assert_expected_output_files(
    expected_output_files: &[&'static str],
    rustc_invocation: Fn())
{
    // or somehow set output directory
    run_in_tmpdir(|| {
        rustc_invocation();
        let actual_output_files = find_files_in_directory(); // pseudocode
        assert_eq!(&actual_output_files, expected_output_files, "output artifacts mismatch");
    });
}
```

File: tests/run-make/output-type-permutations/rmake.rs [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+// Each test takes 4 arguments:
+// `must_exist`: output files which must be found - if any are absent, the test fails
+// `can_exist`: optional output files which will not trigger a failure
+// `dir`: the name of the directory where the test happens
+// `rustc_invocation`: the rustc command being tested
+// Any unexpected output files not listed in `must_exist` or `can_exist` will cause a failure.
+fn assert_expected_output_files(
+    must_exist: &[&'static str],
+    can_exist: &[&'static str],
+    dir: &str,
```
Comment (jieyouxu) [inline tests/run-make/output-type-permutations/rmake.rs:21]: Suggestion: so I thought maybe we could have a param struct for `must_exist`, `can_exist` and `dir`, something like

```rs
struct Expectations {
    /// Name of the directory where the test happens.
    test_dir: String,
    /// Output files which must be found. The test fails if any are absent.
    expected_files: Vec<String>,
    /// Allowed output files which will not trigger a failure.
    allowed_files: Vec<String>,
}
```

and callsite (each test call) can be like

```rs
assert_expected_output_files(
    Expectations {
        test_dir: "foo.bar".to_string(),
        expected_files: s!["a.lib", static_lib_name("bar")],
        allowed_files: s!["b.dylib"],
    },
    || { todo!() }
);
```

I find this helpful because it clearly documents which each parameter means local to each `assert_expected_output_files` call.

the `s![]` macro is just a little helper so we can have a mixed list of various kinds of strings (so as long as they impl `ToString`), it just converts them to a simple `Vec<String>`.

```rs
macro_rules! s {
    ( $( $x:expr ),* ) => {
        {
            let mut temp_vec = Vec::new();
            $(
                temp_vec.push($x.to_string());
            )*
            temp_vec
        }
    };
}
```

This avoids having to do the `Box::leak(Box::new(...))` dance you have below.

**Review by jieyouxu — COMMENTED** [review]
Thanks, I do find the new structure easier to follow because each sub-test is cleanly grouped. I have some more suggestions which I think makes the test easier to read.

File: tests/run-make/output-type-permutations/rmake.rs (hunk 7 above)
Comment (jieyouxu) [reply tests/run-make/output-type-permutations/rmake.rs:21]: Note that I used a `s![]` declarative macro here and not try to do something like `I: IntoIterator, I::Item: AsRef<str>` because that can't handle heterogeneous types (like having `[&'a str, String]` wouldn't work.

Comment (Oneirical) [reply tests/run-make/output-type-permutations/rmake.rs]: Implemented, see latest commit.

**Review by jieyouxu — APPROVED** [review]
Cool, let's run some try jobs, r=me if they come back green.

**Comment by Oneirical** [comment]
Just a type, wrote `ddl` instead of `dll`.

**Comment by Oneirical** [comment]
@rustbot review
I don't have delegation to try again.
@bors try

**Comment by jieyouxu** [comment]
Accidentally r+'d, oops
@bors try

**Comment by Oneirical** [comment]
`foo.pdb` was missing in the allow-list of one of the rustc calls.

@bors try

File: src/tools/run-make-support/src/lib.rs [hunk 8]
```diff
@@ -261,6 +261,42 @@ pub fn test_while_readonly<P: AsRef<Path>, F: FnOnce() + std::panic::UnwindSafe>
...
+    path.as_ref().file_name().is_some_and(|name| name.to_str().unwrap().starts_with(prefix))
+}
+
+/// Returns true if the filename at `path` has the extension `extension`.
+pub fn has_extension<P: AsRef<Path>>(path: P, extension: &str) -> bool {
+    path.as_ref().extension().is_some_and(|ext| ext == extension)
+}
+
+/// Returns true if the filename at `path` is not in `expected`.
+pub fn filename_not_in_denylist<P: AsRef<Path>>(path: P, expected: &Vec<String>) -> bool {
```
Comment (Kobzol) [inline src/tools/run-make-support/src/lib.rs:294]: Drive-by comment: this can be just `&[String]` (in general, passing `&Vec<X>` anywhere usually isn't needed, and `&[X]` is more general.

File: tests/run-make/output-type-permutations/rmake.rs [hunk 9]
```diff
@@ -0,0 +1,541 @@
...
+// Any unexpected output files not listed in `must_exist` or `can_exist` will cause a failure.
+fn assert_expected_output_files(expectations: Expectations, rustc_invocation: impl Fn()) {
+    let must_exist = expectations.expected_files;
+    let can_exist = expectations.allowed_files;
+    let dir = expectations.test_dir;
+
+    fs_wrapper::create_dir(&dir);
+    rustc_invocation();
+    for file in must_exist {
+        fs_wrapper::remove_file(dir.to_owned() + "/" + &file);
```
Comment (Kobzol) [inline tests/run-make/output-type-permutations/rmake.rs:26]: `test_dir` could be `PathBuf`, when you're anyway treating it as a path here. Who knows how `/` will behave on Windows (although it probably should handle it).

Comment (Oneirical) [reply tests/run-make/output-type-permutations/rmake.rs:26]: > Who knows how / will behave on Windows

It should have no effect, since `fs_wrapper::remove_file` takes an `AsRef<Path>`.

Comment (Kobzol) [reply tests/run-make/output-type-permutations/rmake.rs:26]: That doesn't automatically make it work, e.g. on Linux with backwards slashes it wouldn't work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1eb613878a6de3ba1eceac811eca2211

But hopefully Windows will be able to deal with forward slashes :)

Comment (Oneirical) [reply tests/run-make/output-type-permutations/rmake.rs:26]: That's interesting. Well, let's see the output of the msvc try-job, it should work this time. I hope.

Comment (jieyouxu) [reply tests/run-make/output-type-permutations/rmake.rs:26]: > `test_dir` could be `PathBuf`, when you're anyway treating it as a path here. Who knows how `/` will behave on Windows (although it probably should handle it).

Yeah please pass a PathBuf here and use join which is *designed* to properly handle platform separators, please don't use string concat here with `/`.

**Comment by Oneirical** [comment]
I added the `PathBuf` on the line `fs_wrapper::remove_file(PathBuf::from(&dir).join(&file));` and the Windows test passes. If you think the test is ready, do not r+ just yet as it will conflict with #126709.

@rustbot review

File: tests/run-make/output-type-permutations/rmake.rs [hunk 10]
```diff
@@ -0,0 +1,542 @@
...
+                .crate_type("staticlib")
+                .run();
+        },
+    );
+
+    assert_expected_output_files(
+        Expectations{ expected_files:
+        s!["bar.ll", "bar.s", "bar.o", static_lib_name("bar")],
+        allowed_files: s!["bar.bc"], test_dir: // keep this one for the next test
+        "staticlib-all3".to_string()},
```
Comment (jieyouxu) [inline tests/run-make/output-type-permutations/rmake.rs:509]: Nit: this formatting seems a bit weird?

**Review by jieyouxu — APPROVED** [review]
Thanks, looks good, can you x fmt this test? (does x fmt currently work on rmake.rs, I forgor)

**Comment by jieyouxu** [comment]
Marking this as blocked

@rustbot blocked (waiting on https://github.com/rust-lang/rust/pull/126709)

File: tests/run-make/output-type-permutations/rmake.rs (hunk 10 above)
Comment (Oneirical) [reply tests/run-make/output-type-permutations/rmake.rs]: Just fixed it, I think the comment was messing with the auto-format.
